### PR TITLE
make Glyph.name optional for compatibilty with defcon

### PR DESCRIPTION
**Pull request by anthrotype**
as @typemytype wrote in https://github.com/googlefonts/ufo2ft/issues/363#issuecomment-588356886

> Its maybe not a good plan to make the api between the ufoLib2 and defcon so different. According the spec a glyph name is not required, so on init of a glyph object the name should not be a requirement. But ufoLib2 makes it a required argument, I think this is wrong... and makes packages like ufo2ft complex where it should not be...

this PR makes Glyph.name default to None (e.g. for Glyphs without a parent layer).
As soon as a name-less glyph is added to a Layer object, it gets a name defined by its key in the layer underlying dict.

### CODE REVIEW

File: tests/objects/test_layer.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    layer = Layer(glyphs={"a": g})
+    assert g.name == "a"
+
+    layer.renameGlyph("a", "a")  # no-op
+    assert g.name == "a"
+
+    layer.renameGlyph("a", "b")
+    assert g.name == "b"
+
+    layer.insertGlyph(g, "a")
```
Comment (madig) [inline tests/objects/test_layer.py]: I wonder if we should really allow N:1 mappings of names to objects. Doesn't this lead to funny corner cases that we can't debug?

File: src/ufoLib2/objects/layer.py [hunk 2]
```diff
@@ -123,21 +157,25 @@ def loadGlyph(self, name):
...
+            raise KeyError(f"target glyph named '{newName}' already exists")
         # pop and set name
         glyph = self.pop(name)
         glyph._name = newName
         # add it back
         self._glyphs[newName] = glyph
 
+    def instantiateGlyphObject(self):
+        # only for defcon API compatibility
+        return Glyph()
```
[the remark this answers is not in the document]
Comment (anthrotype) [reply src/ufoLib2/objects/layer.py]: no, it doesn't
https://github.com/robotools/defcon/blob/71f1cd9071c284e470103a54bc1f6d50375f85c8/Lib/defcon/objects/layer.py#L133-L144

File: tests/objects/test_layer.py (hunk 1 above)
Comment (anthrotype) [reply tests/objects/test_layer.py]: look closer, insertGlyph copies by default

**Comment by typemytype** [comment]
thanks!

**Comment by anthrotype** [comment]
@typemytype do let us know if you find other API incompatibilities that we could fix, thanks
